docs/site/Express-middleware.md

```diff
@@ -0,0 +1,319 @@
+implementation. We decided to not expose middleware capabilities to users while
+we pursue an elegant and non-invasive way to fit Express middleware into the
+LoopBack 4 programming model nicely. Meanwhile, we have received various
+requests and questions from our users on how to use Express middleware with
+LoopBack 4 or migrate their usage of Express middleware from LoopBack 3 to
+LoopBack 4.
+
+## Use cases
+
+The following use cases are identified to allow Express middleware to work with
+LoopBack 4:
+
+### For application developers
+
+1. Invoke one or more Express middleware handler functions in the REST sequence
+   before, after, or between existing actions.
+2. Apply Express middleware globally to all controllers, or locally to certain
+   controller classes and methods.
+3. Migrate LoopBack 3 applications that use Express middleware
+4. Be able to separate Express middleware configuration from its registration
+   and allow configuration changes at runtime to be effective without restarting
+   the application.
+
+### For extension developers
+
+1. Allow new actions to be added to the REST sequence without requiring code
+   changes to the application as long as the extension component is mounted to
+   the application. For example, `@loopback/authentication` contributes an
+   `Authenticate` action. It should be possible to build an extension module for
+   [`Helmet`](https://helmetjs.github.io/) to provide better protection for
+   LoopBack.
+2. Allow actions to be invoked in a configurable order.
+3. Allow extensions to leverage LoopBack's dependency injection and extension
+   point/extension capabilities.
+
+## Bring Express middleware to LoopBack
+
+Let's start with a few examples to illustrate how we can bring Express
+middleware to LoopBack applications with minimal efforts.
+
+### Use Express middleware within the sequence of actions
+
+Express middleware can now be plugged into the REST sequence with an
+`InvokeMiddleware` action being injected to the default sequence.
+
+#### Invoke Express middleware explicitly in the sequence
+
+The custom sequence class below invokes two Express middleware
+([`helmet`](https://helmetjs.github.io/) and
+[`morgan`](https://github.com/expressjs/morgan)) handler functions as the first
+step.
+
+{% include code-caption.html content="src/sequence.ts" %}
+
+```ts
+import helmet from 'helmet'; // For security
+import morgan from 'morgan'; // For http access logging
+
+export class MySequence extends DefaultSequence {
+  async handle(context: RequestContext): Promise<void> {
+    try {
+      const {request, response} = context;
+      // `this.invokeMiddleware` is an injected function to invoke a list of
+      // Express middleware handler functions
+      const finished = await this.invokeMiddleware(context, [
+        helmet({}), // options for helmet is fixed and cannot be changed at runtime
+        morgan('combined', {}), // options for morgan is fixed and cannot be changed at runtime
```

Comment:
Please note we must not call middleware factories for every incoming request, that would be very bad for performance.

Proposed change:

```ts
const middleware = [
  helmet({}), // options for helmet is fixed and cannot be changed at runtime
  morgan('combined', {}), // options for morgan is fixed and cannot be changed at runtime
];

export class MySequence extends DefaultSequence {
   async handle(context: RequestContext): Promise<void> {
      // ...
      const finished = await this.invokeMiddleware(context, middleware);
      // ...
  }
}
```